### PR TITLE
Redesign tag helper codegen

### PR DESCRIPTION
**Pull request by rynowak**

### CODE REVIEW

**Comment by rynowak** [comment]
/cc @NTaylorMullen 

I know there are a few test failure right now, but my VS can't run tests anymore 👎 . I'll fix them in the morning.

**Comment by rynowak** [comment]
I was really pleased how little diff there ends up being in the codegen.

File: src/Microsoft.AspNetCore.Mvc.Razor.Extensions/ViewComponentTagHelperPass.cs [hunk 1]
```diff
@@ -1,106 +1,168 @@
...
 using System.Linq;
 using System.Threading.Tasks;
 using Microsoft.AspNetCore.Razor.Language;
 using Microsoft.AspNetCore.Razor.Language.CodeGeneration;
+using Microsoft.AspNetCore.Razor.Language.Extensions;
 using Microsoft.AspNetCore.Razor.Language.Intermediate;
 
 namespace Microsoft.AspNetCore.Mvc.Razor.Extensions
 {
     public class ViewComponentTagHelperPass : IntermediateNodePassBase, IRazorOptimizationPass
```
Comment (ajaybhargavb) [inline src/Microsoft.AspNetCore.Mvc.Razor.Extensions/ViewComponentTagHelperPass.cs]: Should this be called `ViewComponentTagHelperOptimizationPass`?

File: src/Microsoft.AspNetCore.Mvc.Razor.Extensions/ViewComponentTagHelperPass.cs [hunk 2]
```diff
@@ -1,106 +1,168 @@
...
+            while (i < context.Class.Children.Count && context.Class.Children[i] is FieldDeclarationIntermediateNode)
+            {
+                i++;
+            }
+
+            context.Class.Children.Insert(i, new FieldDeclarationIntermediateNode()
+            {
+                Annotations =
+                {
+                    { "TagHelperField", bool.TrueString },
```
Comment (ajaybhargavb) [inline src/Microsoft.AspNetCore.Mvc.Razor.Extensions/ViewComponentTagHelperPass.cs]: Put this in `CommonAnnotations`?

File: src/Microsoft.AspNetCore.Razor.Language/Extensions/DefaultTagHelperHtmlAttributeIntermediateNode.cs [hunk 3]
```diff
@@ -0,0 +1,76 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using Microsoft.AspNetCore.Razor.Language.CodeGeneration;
+using Microsoft.AspNetCore.Razor.Language.Intermediate;
+
+namespace Microsoft.AspNetCore.Razor.Language.Extensions
+{
+    public class DefaultTagHelperHtmlAttributeIntermediateNode : ExtensionIntermediateNode
```
Comment (ajaybhargavb) [inline src/Microsoft.AspNetCore.Razor.Language/Extensions/DefaultTagHelperHtmlAttributeIntermediateNode.cs]: sealed?

File: src/Microsoft.AspNetCore.Razor.Language/Extensions/DefaultTagHelperTargetExtension.cs [hunk 4]
```diff
@@ -0,0 +1,574 @@
...
+        {
+            var propertyAccessor = $"{tagHelperVariableName}.{descriptor.GetPropertyName()}";
+
+            if (isIndexerNameMatch)
+            {
+                var dictionaryKey = attributeName.Substring(descriptor.IndexerNamePrefix.Length);
+                propertyAccessor += $"[\"{dictionaryKey}\"]";
+            }
+
+            return propertyAccessor;
```
Comment (ajaybhargavb) [inline src/Microsoft.AspNetCore.Razor.Language/Extensions/DefaultTagHelperTargetExtension.cs]: I assume this whole file is just copy-paste. Not looking at this.

File: src/Microsoft.AspNetCore.Razor.Language/RazorEngine.cs [hunk 5]
```diff
@@ -104,7 +105,12 @@ internal static void AddDefaults(IRazorEngineBuilder builder)
 
         internal static void AddRuntimeDefaults(IRazorEngineBuilder builder)
         {
+            // Intermediate Node Passes
             builder.Features.Add(new PreallocatedTagHelperAttributeOptimizationPass());
+
+            // DesignTime Runtime Targets
```
Comment (ajaybhargavb) [inline src/Microsoft.AspNetCore.Razor.Language/RazorEngine.cs]: Not designtime

File: src/Microsoft.AspNetCore.Razor.Language/TagHelperDescriptorExtensions.cs [hunk 6]
```diff
@@ -7,16 +7,36 @@ namespace Microsoft.AspNetCore.Razor.Language
...
-            return string.Equals(descriptor.Kind, TagHelperConventions.DefaultKind, StringComparison.Ordinal);
+            if (tagHelper == null)
+            {
+                throw new ArgumentNullException(nameof(tagHelper));
+            }
+
+            return string.Equals(tagHelper.Kind, TagHelperConventions.DefaultKind, StringComparison.Ordinal);
+        }
+
+        public static bool KindUsesDefaultTagHelperRuntime(this TagHelperDescriptor tagHelper)
```
Comment (ajaybhargavb) [inline src/Microsoft.AspNetCore.Razor.Language/TagHelperDescriptorExtensions.cs]: I am confused about what this method is for. I understand the need for `IsDefaultKind()`. Maybe I am just confused by the word "runtime".

File: test/Microsoft.AspNetCore.Mvc.Razor.Extensions.Test/TestFiles/IntegrationTests/CodeGenerationIntegrationTest/ModelExpressionTagHelper_Runtime.ir.txt [hunk 7]
```diff
@@ -27,14 +28,15 @@ Document -
                 CSharpCode - 
                     IntermediateToken -  - CSharp - BeginContext(66, 25, false);
                 TagHelper - (66:4,0 [25] ModelExpressionTagHelper.cshtml) - input-test - TagMode.SelfClosing
-                    TagHelperBody - 
-                    CreateTagHelper -  - InputTestTagHelper
-                    SetTagHelperProperty - (83:4,17 [4] ModelExpressionTagHelper.cshtml) - for - For - HtmlAttributeValueStyle.DoubleQuotes
+                    DefaultTagHelperBody - 
+                    DefaultTagHelperCreate -  - InputTestTagHelper
+                    DefaultTagHelperProperty - (83:4,17 [4] ModelExpressionTagHelper.cshtml) - for - Microsoft.AspNetCore.Mvc.ViewFeatures.ModelExpression InputTestTagHelper.For - HtmlAttributeValueStyle.DoubleQuotes
```
Comment (ajaybhargavb) [inline test/Microsoft.AspNetCore.Mvc.Razor.Extensions.Test/TestFiles/IntegrationTests/CodeGenerationIntegrationTest/ModelExpressionTagHelper_Runtime.ir.txt]: I see some inconsistencies where some places we use `DefaultTagHelperProperty` and other we use `DefaultTagHelperPropertyValue`. We should make them all be the same.

File: test/Microsoft.AspNetCore.Mvc.Razor.Extensions.Test/TestFiles/IntegrationTests/CodeGenerationIntegrationTest/RazorPagesWithoutModel_DesignTime.ir.txt [hunk 8]
```diff
@@ -49,15 +50,16 @@ Document -
                     IntermediateToken - (449:21,43 [1] RazorPagesWithoutModel.cshtml) - Html - >
                     IntermediateToken - (450:21,44 [6] RazorPagesWithoutModel.cshtml) - Html - \n    
                 TagHelper - (456:22,4 [31] RazorPagesWithoutModel.cshtml) - div - TagMode.StartTagAndEndTag
-                    TagHelperBody - 
-                    CreateTagHelper -  - DivTagHelper
-                    AddTagHelperHtmlAttribute -  - class - AttributeStructure.DoubleQuotes
+                    DefaultTagHelperBody - 
+                    DefaultTagHelperCreate -  - DivTagHelper
+                    DefaultTagHelperHtmlAttribute -  - class - HtmlAttributeValueStyle.DoubleQuotes
```
Comment (ajaybhargavb) [inline test/Microsoft.AspNetCore.Mvc.Razor.Extensions.Test/TestFiles/IntegrationTests/CodeGenerationIntegrationTest/RazorPagesWithoutModel_DesignTime.ir.txt]: Why did this change? `AttributeStructure` => `HtmlAttributeValueStyle`

File: test/Microsoft.AspNetCore.Mvc.Razor.Extensions.Test/TestFiles/IntegrationTests/CodeGenerationIntegrationTest/ViewComponentTagHelper_DesignTime.codegen.cs [hunk 9]
```diff
@@ -14,6 +14,8 @@ namespace AspNetCore
     using Microsoft.AspNetCore.Mvc.ViewFeatures;
     public class TestFiles_IntegrationTests_CodeGenerationIntegrationTest_ViewComponentTagHelper_cshtml : global::Microsoft.AspNetCore.Mvc.Razor.RazorPage<dynamic>
     {
+        private global::AllTagHelper __AllTagHelper;
+        private global::AspNetCore.TestFiles_IntegrationTests_CodeGenerationIntegrationTest_ViewComponentTagHelper_cshtml.__Generated__TestViewComponentTagHelper __TestViewComponentTagHelper;
```
Comment (ajaybhargavb) [inline test/Microsoft.AspNetCore.Mvc.Razor.Extensions.Test/TestFiles/IntegrationTests/CodeGenerationIntegrationTest/ViewComponentTagHelper_DesignTime.codegen.cs]: The field name doesn't include the fullname like we had before. Was this intentional?

File: test/Microsoft.AspNetCore.Razor.Language.Test/Microsoft.AspNetCore.Razor.Language.Test.csproj [hunk 10]
```diff
@@ -10,6 +10,7 @@
 
   <ItemGroup>
     <EmbeddedResource Include="TestFiles\**\*" />
+    <Compile Remove="CodeGeneration\DesignTimeTagHelperWriterTest.cs" />
```
Comment (ajaybhargavb) [inline test/Microsoft.AspNetCore.Razor.Language.Test/Microsoft.AspNetCore.Razor.Language.Test.csproj]: Assuming this is just temporary.

File: src/Microsoft.AspNetCore.Razor.Language/RazorDiagnosticFactory.cs [hunk 11]
```diff
@@ -66,7 +66,9 @@ public static RazorDiagnostic CreateParsing_DuplicateDirective(string directive,
                 $"{DiagnosticPrefix}3000",
                 () => Resources.TagHelper_InvalidRestrictedChildNullOrWhitespace,
                 RazorDiagnosticSeverity.Error);
+
         public static RazorDiagnostic CreateTagHelper_InvalidRestrictedChildNullOrWhitespace(string tagHelperDisplayName)
+
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Razor.Language/RazorDiagnosticFactory.cs]: 😱

File: test/Microsoft.AspNetCore.Mvc.Razor.Extensions.Test/TagHelperDescriptorExtensionsTest.cs [hunk 12]
```diff
@@ -11,7 +11,6 @@ public class TagHelperDescriptorExtensionsTest
         [Fact]
         public void IsViewComponentKind_ReturnsFalse_ForNonVCTHDescriptor()
         {
-            //Arrange
```
Comment (NTaylorMullen) [inline test/Microsoft.AspNetCore.Mvc.Razor.Extensions.Test/TagHelperDescriptorExtensionsTest.cs]: ?

File: test/Microsoft.AspNetCore.Razor.Language.Test/TestFiles/IntegrationTests/CodeGenerationIntegrationTest/BasicTagHelpers_DesignTime.codegen.cs [hunk 13]
```diff
@@ -14,20 +17,12 @@ public class TestFiles_IntegrationTests_CodeGenerationIntegrationTest_BasicTagHe
...
-        private global::TestNamespace.PTagHelper __TestNamespace_PTagHelper = null;
-        private global::TestNamespace.InputTagHelper __TestNamespace_InputTagHelper = null;
-        private global::TestNamespace.InputTagHelper2 __TestNamespace_InputTagHelper2 = null;
         #pragma warning disable 1998
         public async System.Threading.Tasks.Task ExecuteAsync()
         {
             __TestNamespace_PTagHelper = CreateTagHelper<global::TestNamespace.PTagHelper>();
             __TestNamespace_InputTagHelper = CreateTagHelper<global::TestNamespace.InputTagHelper>();
             __TestNamespace_InputTagHelper2 = CreateTagHelper<global::TestNamespace.InputTagHelper2>();
-#line 6 "TestFiles/IntegrationTests/CodeGenerationIntegrationTest/BasicTagHelpers.cshtml"
```
Comment (NTaylorMullen) [inline test/Microsoft.AspNetCore.Razor.Language.Test/TestFiles/IntegrationTests/CodeGenerationIntegrationTest/BasicTagHelpers_DesignTime.codegen.cs]: This doesn't look right.

File: test/Microsoft.AspNetCore.Razor.Language.Test/TestFiles/IntegrationTests/CodeGenerationIntegrationTest/ComplexTagHelpers_DesignTime.codegen.cs [hunk 14]
```diff
@@ -110,11 +110,6 @@ public class TestFiles_IntegrationTests_CodeGenerationIntegrationTest_ComplexTag
 #line default
 #line hidden
             __TestNamespace_PTagHelper = CreateTagHelper<global::TestNamespace.PTagHelper>();
-#line 9 "TestFiles/IntegrationTests/CodeGenerationIntegrationTest/ComplexTagHelpers.cshtml"
```
Comment (NTaylorMullen) [inline test/Microsoft.AspNetCore.Razor.Language.Test/TestFiles/IntegrationTests/CodeGenerationIntegrationTest/ComplexTagHelpers_DesignTime.codegen.cs]: Another one. Wont comment on anymore.

File: src/Microsoft.AspNetCore.Mvc.Razor.Extensions/ViewComponentTagHelperPass.cs (hunk 1 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Razor.Extensions/ViewComponentTagHelperPass.cs]: meh

File: src/Microsoft.AspNetCore.Mvc.Razor.Extensions/ViewComponentTagHelperPass.cs (hunk 2 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Razor.Extensions/ViewComponentTagHelperPass.cs]: Yeah!

File: src/Microsoft.AspNetCore.Razor.Language/Extensions/DefaultTagHelperHtmlAttributeIntermediateNode.cs (hunk 3 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Razor.Language/Extensions/DefaultTagHelperHtmlAttributeIntermediateNode.cs]: yes, good catch

File: src/Microsoft.AspNetCore.Razor.Language/Extensions/DefaultTagHelperTargetExtension.cs (hunk 4 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Razor.Language/Extensions/DefaultTagHelperTargetExtension.cs]: Yeah mostly

File: src/Microsoft.AspNetCore.Razor.Language/RazorEngine.cs (hunk 5 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Razor.Language/RazorEngine.cs]: k

File: src/Microsoft.AspNetCore.Razor.Language/TagHelperDescriptorExtensions.cs (hunk 6 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Razor.Language/TagHelperDescriptorExtensions.cs]: It means that this tag helper uses things like `TagHelperExecutionContext`. Basically everything that uses the `ITagHelper` and related types.

File: test/Microsoft.AspNetCore.Mvc.Razor.Extensions.Test/TestFiles/IntegrationTests/CodeGenerationIntegrationTest/ModelExpressionTagHelper_Runtime.ir.txt (hunk 7 above)
Comment (rynowak) [inline test/Microsoft.AspNetCore.Mvc.Razor.Extensions.Test/TestFiles/IntegrationTests/CodeGenerationIntegrationTest/ModelExpressionTagHelper_Runtime.ir.txt]: I don't see `DefaultTagHelperPropertyValue` anywhere. I don't understand this feedback

File: test/Microsoft.AspNetCore.Mvc.Razor.Extensions.Test/TestFiles/IntegrationTests/CodeGenerationIntegrationTest/RazorPagesWithoutModel_DesignTime.ir.txt (hunk 8 above)
Comment (rynowak) [inline test/Microsoft.AspNetCore.Mvc.Razor.Extensions.Test/TestFiles/IntegrationTests/CodeGenerationIntegrationTest/RazorPagesWithoutModel_DesignTime.ir.txt]: It's just cosmetic. I made the IR serialization consistent. Some nodes were using one, some the other, but they mean the same thing

File: test/Microsoft.AspNetCore.Mvc.Razor.Extensions.Test/TestFiles/IntegrationTests/CodeGenerationIntegrationTest/ViewComponentTagHelper_DesignTime.codegen.cs (hunk 9 above)
Comment (rynowak) [inline test/Microsoft.AspNetCore.Mvc.Razor.Extensions.Test/TestFiles/IntegrationTests/CodeGenerationIntegrationTest/ViewComponentTagHelper_DesignTime.codegen.cs]: Yes, I simplified the field names for VCTH

File: test/Microsoft.AspNetCore.Razor.Language.Test/Microsoft.AspNetCore.Razor.Language.Test.csproj (hunk 10 above)
Comment (rynowak) [inline test/Microsoft.AspNetCore.Razor.Language.Test/Microsoft.AspNetCore.Razor.Language.Test.csproj]: uhhhhh

File: src/Microsoft.AspNetCore.Mvc.Razor.Extensions/ViewComponentTagHelperPass.cs [hunk 15]
```diff
@@ -1,106 +1,168 @@
...
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Razor.Extensions/ViewComponentTagHelperPass.cs]: Fixed

File: src/Microsoft.AspNetCore.Razor.Language/Extensions/DefaultTagHelperOptimizationPass.cs [hunk 16]
```diff
@@ -0,0 +1,259 @@
...
+            {
+                i++;
+            }
+            while (i < node.Children.Count && node.Children[i] is DefaultTagHelperBodyIntermediateNode)
+            {
+                i++;
+            }
+
+            // Now find the last create node.
+            while (i < node.Children.Count && node.Children[i] is DefaultTagHelperCreateIntermediateNode)
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Razor.Language/Extensions/DefaultTagHelperOptimizationPass.cs]: I'm being defensive. If I didn't do it here and did it in VCTH you would make the opposite comment

File: test/Microsoft.AspNetCore.Mvc.Razor.Extensions.Test/TestFiles/IntegrationTests/CodeGenerationIntegrationTest/ModelExpressionTagHelper_Runtime.ir.txt (hunk 7 above)
Comment (ajaybhargavb) [inline test/Microsoft.AspNetCore.Mvc.Razor.Extensions.Test/TestFiles/IntegrationTests/CodeGenerationIntegrationTest/ModelExpressionTagHelper_Runtime.ir.txt]: Typed the wrong thing. I meant all the preallocated stuff has TagHelperPropertyValue in the name

File: test/Microsoft.AspNetCore.Razor.Language.Test/TestFiles/IntegrationTests/CodeGenerationIntegrationTest/BasicTagHelpers_DesignTime.codegen.cs (hunk 13 above)
Comment (rynowak) [inline test/Microsoft.AspNetCore.Razor.Language.Test/TestFiles/IntegrationTests/CodeGenerationIntegrationTest/BasicTagHelpers_DesignTime.codegen.cs]: Fixed

**Comment by rynowak** [comment]
Updated

**Pull request merged**